crates/ruff/src/rules/ruff/rules/explicit_f_string_type_conversion.rs

```diff
@@ -41,72 +46,113 @@ impl AlwaysAutofixableViolation for ExplicitFStringTypeConversion {
-    };
+    Ok(Fix::automatic(Edit::range_replacement(
+        state.to_string(),
+        range,
+    )))
+}
 
-    if !matches!(id.as_str(), "str" | "repr" | "ascii") {
-        return;
-    };
+/// RUF010
+pub(crate) fn explicit_f_string_type_conversion(
+    checker: &mut Checker,
+    expr: &Expr,
+    values: &[Expr],
+) {
+    let mut formatted_values: Vec<usize> = vec![];
+    for (index, value) in values.iter().enumerate() {
+        let ExprKind::FormattedValue(ast::ExprFormattedValue {
+            conversion,
+            value,
+            ..
+        }) = &value.node else {
+            continue;
+        };
+        // Skip if there's already a conversion flag.
+        if *conversion != ast::ConversionFlag::None as u32 {
+            return;
+        }
 
-    if !checker.ctx.is_builtin(id) {
-        return;
-    }
+        let ExprKind::Call(ast::ExprCall {
+            func,
+            args,
+            keywords,
+        }) = &value.node else {
+            return;
+        };
 
-    let mut diagnostic = Diagnostic::new(ExplicitFStringTypeConversion, formatted_value.range());
+        // Can't be a conversion otherwise.
+        if args.len() != 1 || !keywords.is_empty() {
+            return;
+        }
 
-    if checker.patch(diagnostic.kind.rule()) {
-        // Replace the call node with its argument and a conversion flag.
-        let mut conv_expr = expr.clone();
-        let ExprKind::FormattedValue(ast::ExprFormattedValue {
-            ref mut conversion,
-            ref mut value,
-            ..
-        }) = conv_expr.node else {
+        let ExprKind::Name(ast::ExprName { id, .. }) = &func.node else {
             return;
         };
 
-        *conversion = match id.as_str() {
-            "ascii" => ast::Int::new(ast::ConversionFlag::Ascii as u32),
-            "str" => ast::Int::new(ast::ConversionFlag::Str as u32),
-            "repr" => ast::Int::new(ast::ConversionFlag::Repr as u32),
-            &_ => unreachable!(),
+        if !matches!(id.as_str(), "str" | "repr" | "ascii") {
+            return;
         };
 
-        value.node = args[0].node.clone();
+        if !checker.ctx.is_builtin(id) {
+            return;
+        }
+        formatted_values.push(index);
+    }
 
-        diagnostic.set_fix(Fix::automatic(Edit::range_replacement(
-            unparse_expr(&conv_expr, checker.stylist),
-            formatted_value
-                .range()
-                .sub_start(TextSize::from(1))
-                .add_end(TextSize::from(1)),
-        )));
+    if formatted_values.is_empty() {
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(ExplicitFStringTypeConversion, expr.range());
+
+    if checker.patch(diagnostic.kind.rule()) {
+        diagnostic.try_set_fix(|| {
+            fix_explicit_f_string_type_conversion(
+                expr,
+                &formatted_values,
+                checker.locator,
+                checker.stylist,
+            )
+        });
     }
```

Comment:
It looks like we now generate one diagnostic for the whole f-string (and that the diagnostic range is that of the entire f-string).

I think it'd be preferable to retain one diagnostic per formatted value, and limit the diagnostic to the formatted value. It's fine if the autofix contents include the whole string, even for each formatted value, but they should be separate diagnostics. (E.g., we could iterate over `formatted_values` and create one diagnostic per index, with a fix specific to that index.)